app/forms/user_invite_form.rb

```diff
@@ -0,0 +1,71 @@
+class UserInviteForm
+  FORM_PARAMS = %i[first_name last_name email].freeze
+  include ActiveModel::Model
+
+  attr_accessor :email, :first_name, :last_name, :service, :organisation
+
+  validate :validate_user
+  validate :validate_membership
+  validates :service, :organisation, presence: true
+
+  def invite
+    return false unless valid?
+
+    send_invite_email if save_user
+  end
+
+  def as_form_params
+    { "user_invite_form" => slice(FORM_PARAMS) }
+  end
+
+  private
+
+  def user
+    @user ||=
+      begin
+        user = user_klass.find_or_initialize_by(email:)
+        user.assign_attributes(first_name:, last_name:)
+        user
+      end
+  end
+
+  def save_user
+    ActiveRecord::Base.transaction do
+      user.save!
+      membership.save!
+    end
+  end
+
+  def validate_user
+    if user.invalid?
+      user.errors.each do |err|
+        errors.add(err.attribute, err.message)
+      end
+    end
+  end
+
+  def validate_membership
+    if membership.invalid?
+      errors.add(:email, :taken)
+    end
+  end
+
+  def membership
+    @membership ||= user.memberships.new(organisation:)
+  end
+
+  def send_invite_email
+    url = Rails.application.routes.url_helpers.sign_in_url(host:)
+    NotifyMailer.send_organisation_invite_email(user, organisation, url).deliver_later
+  end
```

Comment:
I don't think we should be adding business logic to form objects, right? The way I understand form objects is they are a way to apply data modelling and validations, like models, to forms that do not map 1-1 from model to view.

What do you think?

Reply:
But this is a form for inviting a user, and kicking off an email is essential to inviting. So I think it is the most sensible place for it. Otherwise we just call a mailer from the controller, definitely the wrong place for business logic.

Reply:
Would that business logic belong to a service object?

Reply:
All that service object would do now is call the mailer. Before it was saving -- now that lives in the form. I think having a service object that just calls a single method is an unhelpful level of abstraction.

Reply:
Added user invite service here https://github.com/DFE-Digital/itt-mentor-services/pull/142/commits/dc405e51987c00c9aabe7d2faeb4b97ed2c2fe0c